sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -2781,49 +2971,67 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    *       [SORTED BY (col_name [ASC|DESC], ...)]
    *       INTO num_buckets BUCKETS
    *     ]
    *     [LOCATION path]
    *     [COMMENT table_comment]
    *     [TBLPROPERTIES (property_name=property_value, ...)]
+   *   partition_clauses:
+   *     [PARTITIONED BY (col_name, transform(col_name), transform(constant, col_name), ...)] |
+   *     [PARTITIONED BY (col2[:] data_type [COMMENT col_comment], ...)]
    * }}}
    */
   override def visitCreateTable(ctx: CreateTableContext): LogicalPlan = withOrigin(ctx) {
     val (table, temp, ifNotExists, external) = visitCreateTableHeader(ctx.createTableHeader)
-    if (external) {
-      operationNotAllowed("CREATE EXTERNAL TABLE ...", ctx)
-    }
-    val schema = Option(ctx.colTypeList()).map(createSchema)
+
+    val columns = Option(ctx.colTypeList()).map(visitColTypeList)
     val provider = Option(ctx.tableProvider).map(_.multipartIdentifier.getText)
-    val (partitioning, bucketSpec, properties, options, location, comment) =
+    val (partTransforms, partCols, bucketSpec, properties, options, location, comment, serdeInfo) =
       visitCreateTableClauses(ctx.createTableClauses())
 
+    if (provider.isDefined && serdeInfo.isDefined) {
+      operationNotAllowed(s"CREATE TABLE ... USING ... ${serdeInfo.get.describe}", ctx)
+    }
+
+    val schema = columns
+        .map(dataCols => StructType(dataCols ++ partCols))
+        .getOrElse(StructType(partCols))
+    val partitioning = partitionExpressions(partTransforms, partCols, ctx)
+
     Option(ctx.query).map(plan) match {
       case Some(_) if temp =>
-        operationNotAllowed("CREATE TEMPORARY TABLE ... USING ... AS query", ctx)
+        operationNotAllowed(
+          "CREATE TEMPORARY TABLE ... AS ..., use CREATE TEMPORARY VIEW instead",
+          ctx)
 
-      case Some(_) if schema.isDefined =>
+      case Some(_) if columns.isDefined =>
         operationNotAllowed(
           "Schema may not be specified in a Create Table As Select (CTAS) statement",
           ctx)
 
+      case Some(_) if partCols.nonEmpty =>
+        // non-reference partition columns are not allowed because schema can't be specified
+        operationNotAllowed(
+          "Partition column types may not be specified in Create Table As Select (CTAS)",
+          ctx)
+
       case Some(query) =>
         CreateTableAsSelectStatement(
           table, query, partitioning, bucketSpec, properties, provider, options, location, comment,
-          writeOptions = Map.empty, ifNotExists = ifNotExists)
+          writeOptions = Map.empty, serdeInfo, external = external, ifNotExists = ifNotExists)
 
       case None if temp =>
-        // CREATE TEMPORARY TABLE ... USING ... is not supported by the catalyst parser.
-        // Use CREATE TEMPORARY VIEW ... USING ... instead.
-        operationNotAllowed("CREATE TEMPORARY TABLE IF NOT EXISTS", ctx)
+        operationNotAllowed("CREATE TEMPORARY TABLE", ctx)
 
       case _ =>
```

Comment:
`schema` is only accessed in this branch. It's better to define the variable only in the branch that needs it.

BTW we can follow the previous code style which looks simpler:
```
val columns = Option(ctx.colTypeList()).map(visitColTypeList).getOrElse(Nil)
...
Option(ctx.query).map(plan) match {
  ...
  case _ =>
    // Note: `PARTITION BY col type` requires partition columns to be distinct from the schema, so we need
    // to include the partition columns here explicitly
    val schema = StructType(columns ++ partCols)
}
```